alf/algorithms/vq_vae.py

```diff
@@ -0,0 +1,164 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Vector Quantized Variational AutoEncoder Algorithm."""
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+import alf
+from alf.algorithms.algorithm import Algorithm
+from alf.data_structures import AlgStep, LossInfo, namedtuple
+
+VqvaeLossInfo = namedtuple(
+    "VqvaeLossInfo", ["quantization", "commitment", "reconstruction"],
+    default_value=())
+
+
+class Vqvae(Algorithm):
+    r"""Vector Quantized Variational AutoEncoder (VQVAE) algorithm, described in:
+
+    ::
+        A van den Oord et al. "Neural Discrete Representation Learning", NeurIPS 2017.
+
+    VQVAE is different from standard VAE mainly in the follows aspects:
+
+    1. Discrete latent is used, instead of continuous latent as in standard VAE.
+    2. Standard VAE uses Gaussian prior and posterior. VQVAE can be viewed as
+       using a determinstic form of posterior, which is a categorical
+       distribution with onehot samples computed by nearest neighbor matching
+       (Eq.1 of the paper). By using a uniform prior, the KL divergence is constant.
+
+
+    """
+
+    def __init__(self,
+                 input_tensor_spec,
+                 num_embeddings,
+                 embedding_dim,
+                 encoder_ctor,
+                 decoder_ctor,
+                 optimizer=None,
+                 commitment_loss_weight=1.0,
+                 debug_summaries=False,
+                 name="Vqvae"):
+        """
+        Args:
+            input_tensor_spec (TensorSpec): the tensor spec of
+                the input.
+            num_embeddings (int): the number of embeddings (size of codebook)
+            embedding_dim (int): the dimensionality of embedding vectors
+            encoder_ctor (Callable): called as ``encoder_ctor(observation_spec)``
+                to construct the encoding ``Network``. The network takes raw observation
+                as input and output the latent representation.
+            decoder_ctor (Callable): called as ``decoder_ctor(latent_spec)`` to
+                construct the decoder.
+            optimizer (Optimzer|None): if provided, it will be used to optimize
+                the parameter of encoder_net, decoder_net and embedding vectors.
+            commitment_loss_weight (float): the weight for commitment loss.
+        """
+        super().__init__(debug_summaries=debug_summaries, name=name)
+
+        self._embedding_dim = embedding_dim
+        self._num_embeddings = num_embeddings
+
+        # [n, d]
+        self._embedding = nn.Embedding(self._num_embeddings,
+                                       self._embedding_dim)
+        self._embedding.weight.data.uniform_(-1 / self._num_embeddings,
+                                             1 / self._num_embeddings)
```

Comment:
Can use ``torch.nn.init.uniform_`` on ``self._embedding.weight`` instead

Reply:
Switched to use ``torch.nn.init.uniform_``